### PR TITLE
use mock compiler during browser tests

### PR DESCRIPTION
**Pull request by yann300**

### CODE REVIEW

File: ci/makeMockCompiler.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    key: key.trim(),
+    source: source,
+    optimization: optimization,
+    missingInputs: missingInputs,
+    result: result
+  }
+}
+
+function replaceSolCompiler () {
+  fs.readFile('./test-browser/mockcompiler/compiler.js', 'utf8', function (error, data) {
```
Comment (chriseth) [inline ci/makeMockCompiler.js]: You could also include the compiler as a literal string here and combine it with the compilation results, i.e. write everything directly into soljson.js.

File: test-browser/mockcompiler/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+results.json
```
Comment (chriseth) [inline test-browser/mockcompiler/.gitignore]: Why not use the global gitignore file?

File: test-browser/mockcompiler/compiler.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    _compileJSONCallback: {},
+    _compileJSON: {}
+  };
+
+  loadJSON('./test-browser/mockcompiler/results.json', function (data) {
+    window.Module.mockData = JSON.parse(data);
+  });
+}
+
+function loadJSON (url, callback) {
```
Comment (chriseth) [inline test-browser/mockcompiler/compiler.js]: The suggestion above would also not require such a request.

File: test-browser/simpleContract.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+};
+
+function runTests (browser, testData) {
+  browser
+    .url('http://127.0.0.1:8080/#version=builtin')
+    .waitForElementVisible('.newFile', 10000)
+  browser.assert.notEqual(testData, null)
+  /*
+  TODO push contracts to the editor and test
+  setEditorContent(browser, testData.testSimpleContract.sources.Untitled, function () {
```
Comment (chriseth) [inline test-browser/simpleContract.js]: In another test I used
```
.clearValue('#input textarea')
.setValue('#input textarea', source)
```

File: test-browser/simpleContract.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
+'use strict';
+var fs = require('fs');
+
+module.exports = {
+  'Simple Contract': function (browser) {
```
Comment (chriseth) [inline test-browser/simpleContract.js]: Isn't it better to use nodejs require for that?
Another solution (perhaps for the next iteration) would be to declare sources inside each test with a special delimiter and scan all test files when running `makeMockCompiler`.

File: test-browser/simpleContract.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+function runTests (browser, testData) {
+  browser
+    .url('http://127.0.0.1:8080/#version=builtin')
+    .waitForElementVisible('.newFile', 10000)
+  browser.assert.notEqual(testData, null)
+  /*
+  TODO push contracts to the editor and test
+  setEditorContent(browser, testData.testSimpleContract.sources.Untitled, function () {
+    getCompiledContracts(browser, function (contracts) {
```
Comment (chriseth) [inline test-browser/simpleContract.js]: I think you need some kind of wait loop here. I'm not sure if my solution in `compiling.js` works, but you could take a look.

**Comment by axic** [comment]
@yann300 this is a great idea, I reckon it will speed up (and fix) the tests once and for all :+1:

File: test-browser/simpleContract.js (hunk 5 above)
Comment (yann300) [inline test-browser/simpleContract.js]: this line is just about declaring the test.

File: test-browser/simpleContract.js (hunk 4 above)
Comment (yann300) [inline test-browser/simpleContract.js]: could point me to the line please? i can't find it

File: ci/makeMockCompiler.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    result: result
+  };
+}
+
+function replaceSolCompiler (results) {
+  fs.readFile('./test-browser/mockcompiler/compiler.js', 'utf8', function (error, data) {
+    if (error) {
+      return console.log(error);
+    }
+    data = data + '\n\nvar mockData = \'' + JSON.stringify(results).replace(/(\\t|\\n|'|\\")/g, ' ') + '\'';
```
Comment (chriseth) [inline ci/makeMockCompiler.js]: What is the purpose of the replacement? I think this modifies the compilation results. Why not just `var mockData = ` + JSON.stringify(results);` (without the quotes) - JSON is valid javascript.

File: ci/makeMockCompiler.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    fs.writeFile('./soljson.js', data, 'utf8', function (error) {
+      if (error) {
+        return console.log(error);
```
Comment (chriseth) [inline ci/makeMockCompiler.js]: What is vital here: Make the script return a failure exit code so that the test will stop.

File: src/app.js [hunk 9]
```diff
@@ -554,6 +554,13 @@ var run = function () {
   });
 
   $.getJSON('https://ethereum.github.io/solc-bin/bin/list.json').done(function (data) {
+    function buildVersion (build) {
```
Comment (chriseth) [inline src/app.js]: Why include this change here?

File: test-browser/simpleContract.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
+'use strict';
+var fs = require('fs');
+
+module.exports = {
+  'Simple Contract': function (browser) {
+    initTestData(function (error, testData) {
+      if (error) {
+        console.log(error);
```
Comment (chriseth) [inline test-browser/simpleContract.js]: Again, this should result in a failure.

File: src/app.js (hunk 9 above)
Comment (yann300) [inline src/app.js]: that function has been removed from a previous commit. but it is used

Comment (axic) [inline src/app.js]: Oh, true! It is used in the option list. **Damn tests**, we need them so badly.

Comment (axic) [inline src/app.js]: Moved this to #275.

**Comment by chriseth** [comment]
This does not yet call the compiler.

**Comment by chriseth** [comment]
@yann300 ok, let's merge this even if it does not have a test yet. Did you get the ballot working? If not, it's fine for now.

File: src/app.js [hunk 11]
```diff
@@ -520,15 +520,7 @@ var run = function () {
     } else {
       url = 'https://ethereum.github.io/solc-bin/bin/' + version;
     }
-    var isFirefox = typeof InstallTrigger !== 'undefined';
```
Comment (axic) [inline src/app.js]: Why is this being removed? Is the URL fix in #277 solving this problem?

Comment (yann300) [inline src/app.js]: I just want to check if running without workers fix all my tests in firefox. #277 seems to not fix the issue

Comment (axic) [inline src/app.js]: Nothing seems to fix that issue and I cannot even reproduce it :(

Comment (axic) [inline src/app.js]: @yann300 true means using worker, so with removing it in every case it is using in the browser thread.

Comment (yann300) [inline src/app.js]: which firefox version are you running?
so https://ethereum.github.io/browser-solidity/#version=builtin
is working for you?

Comment (axic) [inline src/app.js]: No, it isn't: `Worker error: undefined`.

But the other worker issue I cannot reproduce.

Comment (yann300) [inline src/app.js]: I was only referring to this issue `Worker error: undefined`.
@chriseth the build is ok if we don't use the worker for firefox. 
the Ballot is included but not yet tested (see the comment in the code).
We could merge this now and fix the worker in another pr.

Comment (chriseth) [inline src/app.js]: @axic @yann300 let's first discuss and fix the worker issue in #277

**Comment by chriseth** [comment]
Rebased to test the worker.

**Pull request merged**